WallsAndHoles/simpletexturedrenderer.cpp

```diff
@@ -0,0 +1,401 @@
+    // Destroying buffers doesn't seem to require a valid OpenGL context.
+    mVAOs.remove(&obj);
+    mObjectVertexPositions.remove(&obj);
+    mObjectVertexNormals.remove(&obj);
+    mObjectVertexMaterials.remove(&obj);
+    mObjectVertexTexCoords.remove(&obj);
+    mNumVertices.remove(&obj);
+
+    // If the object's image has an associated texture, remove the object
+    // from the texture's set.
+    if (mImagesToTextures.contains(&obj.getImage())) {
+        QSharedPointer<QOpenGLTexture> texturePtr = mImagesToTextures[&obj.getImage()];
+
+        auto &textureUsageSet = mTexturesToObjects[texturePtr.data()];
+
+        textureUsageSet.remove(&obj);
+
+        // If the texture has no users, unload it.
+        if (textureUsageSet.size() <= 0) {
+            // It is important to unlock and relock the mutex here, just in case
+            // callFunctionOnOpenGL() happens in this thread.
+            locker.unlock();
+            callFunctionOnOpenGL(&SimpleTexturedRenderer::removeImageTexture, this, &obj.getImage());
+            locker.relock();
+        }
+    }
+}
+
+
+void SimpleTexturedRenderer::clearAll()
+{
+    QMutexLocker locker(&mGLDataMutex);
+
+    // Destroying buffers does not require a valid OpenGL context. However,
+    // the same is not true for textures.
+    mVAOs.clear();
+    mObjectVertexPositions.clear();
+    mObjectVertexNormals.clear();
+    mObjectVertexMaterials.clear();
+    mObjectVertexTexCoords.clear();
+    mNumVertices.clear();
+
+    // TODO: Is it possible that new images might be added before this happens?
+    callFunctionOnOpenGL(&SimpleTexturedRenderer::clearAllTextures, this);
+}
+
+void SimpleTexturedRenderer::cleanUp()
+{
+    QMutexLocker locker(&mGLDataMutex);
+
+    mVAOs.clear();
+    mObjectVertexPositions.clear();
+    mObjectVertexNormals.clear();
+    mObjectVertexMaterials.clear();
+    mObjectVertexTexCoords.clear();
+    mNumVertices.clear();
+
+    mShaderProgram.destroy();
+
+    // Unlock the data mutex because the clearAllTextures() function uses it.
+    locker.unlock();
+
+    clearAllTextures();
+}
+
+
+void SimpleTexturedRenderer::paint(QMatrix4x4 mvpMatrix, QVector3D camPos)
+{
+    QMutexLocker locker(&mGLDataMutex);
+    glEnable(GL_DEPTH_TEST);
+
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+    // Set program.
+    mShaderProgram.bind();
+
+    // Set the matrix.
+    mShaderProgram.setUniformMVP(mvpMatrix);
+
+    // Set the lighting uniforms.
+    mShaderProgram.setUniformPointToLight(QVector3D(1, 1, 0).normalized());
+    mShaderProgram.setUniformCameraPosition(camPos);
+    mShaderProgram.setUniformAmbientColor(QVector3D(0.1, 0.1, 0.2));
+    mShaderProgram.setUniformSourceDiffuseColor(QVector3D(0.3, 0.3, 0.2));
+    mShaderProgram.setUniformSourceSpecularColor(QVector3D(0.3, 0.3, 0.2));
+
+    // Draw objects by texture group.
+    foreach (const QImage *img, mImagesToTextures.keys()) {
```

Comment:
https://www.dvratil.cz/2015/06/qt-containers-and-c11-range-based-loops/
because it might be inefficient. Then again, according to https://www.kdab.com/goodbye-q_foreach/ it may be getting deprecated.